File: implementations/merge_sort/typescript/MergeSort.ts
export function mergeSort(arr: number[]): number[] {
  const n = arr.length;
  if (n <= 1) {
    return arr;
  }

  const mid = Math.floor(n / 2);
  const left = mergeSort(arr.slice(0, mid));
  const right = mergeSort(arr.slice(mid));

  return merge(left, right);
}

function merge(left: number[], right: number[]): number[] {
  const mergedArr: number[] = [];
  let i = 0;
  let j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      mergedArr.push(left[i]);
      i++;
    } else {
      mergedArr.push(right[j]);
      j++;
    }
  }

  while (i < left.length) {
    mergedArr.push(left[i]);
    i++;
  }

  while (j < right.length) {
    mergedArr.push(right[j]);
    j++;
  }

  return mergedArr;
}

import fs from "fs"

const input = fs.readFileSync(process.argv[2], "utf8")
const arr = input.split(" ").map((item) => parseInt(item))

let init = Date.now()
const sortedArr = mergeSort(arr)
let end = Date.now()
let isCorrect = true
for (let i = 0; i < sortedArr.length - 1; ++i) {
  if (sortedArr[i] > sortedArr[i + 1]) {
    isCorrect = false
    break
  }
}
process.stdout.write("typescript elapsed seconds "+(end-init)/1000+" | correct: "+isCorrect+"\n")

fs.writeFileSync(process.argv[2]+".merge_sort.out.typescript.txt", sortedArr.join(" "))

File: implementations/merge_sort/rust/merge_sort.rs
use std::env;
use std::fs;
use std::time::Instant;
use std::io::{self, Write};

fn merge_sort(arr: &mut [i32]) {
  let len = arr.len();
  if len <= 1 {
      return;
  }
  let mid = len / 2;
  merge_sort(&mut arr[0..mid]);
  merge_sort(&mut arr[mid..len]);
  let mut temp = arr.to_vec();
  let mut i = 0;
  let mut j = mid;
  for k in 0..len {
      if i < mid && (j >= len || arr[i] <= arr[j]) {
          temp[k] = arr[i];
          i += 1;
      } else {
          temp[k] = arr[j];
          j += 1;
      }
  }
  arr.copy_from_slice(&temp);
}

fn main() {
    let file_path = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(&file_path)
        .expect("Error reading the file");

    let mut input: Vec<i32> = contents
        .split(" ")
        .map(|x| x.parse().expect("Not a number!"))
        .collect();

    let now = Instant::now();
    merge_sort(&mut input);
    let elapsed = now.elapsed();
    
    let output: String = input.iter().map( |&entry| entry.to_string() + " ").collect();
    
    fs::write(file_path+".merge_sort.out.rust.txt", output.trim())
        .expect("Unable to write file");

    let mut is_correct = true;
    for i in 0..input.len()-1 {
        if input[i] > input[i+1] {
            is_correct = false;
            break;
        }
    }

    println!("rust elapsed seconds {} | correct: {}", elapsed.as_secs() as f64 + elapsed.subsec_nanos() as f64 * 1e-9, is_correct);
    io::stdout().flush().unwrap();
}
